src/OmniSharp.Roslyn.CSharp/Services/ExternalSourceServiceFactory.cs

```diff
@@ -0,0 +1,77 @@
+﻿using Microsoft.CodeAnalysis.Host;
+using Microsoft.Extensions.Logging;
+using OmniSharp.Options;
+using OmniSharp.Roslyn.CSharp.Services.Decompilation;
+using OmniSharp.Services;
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OmniSharp.Roslyn.CSharp.Services
+{
+    [Export(typeof(ExternalSourceServiceFactory)), Shared]
+    public class ExternalSourceServiceFactory
+    {
+        private MetadataExternalSourceService _metadataExternalSourceService;
+        private DecompilationExternalSourceService _decompilationExternalSourceService;
+        private readonly IAssemblyLoader _assemblyLoader;
+        private readonly ILoggerFactory _loggerFactory;
+        private static object padlock = new object();
+
+        [ImportingConstructor]
+        public ExternalSourceServiceFactory(IAssemblyLoader assemblyLoader, ILoggerFactory loggerFactory)
+        {
+            _assemblyLoader = assemblyLoader;
+            _loggerFactory = loggerFactory;
+        }
+
+        public IExternalSourceService Create(OmniSharpOptions omniSharpOptions, HostLanguageServices hostLanguageServices)
+        {
+            var enableDecompilationSupport = omniSharpOptions.RoslynExtensionsOptions.EnableDecompilationSupport;
+
+            if (enableDecompilationSupport)
+            {
+                if (_decompilationExternalSourceService == null)
```

Comment:
Can we solve this by having a `Lazy<DecompilationExternalSourceService>` instead. Create the delegate in constructor where we have assembly loader and logger factory.

Reply:
I actually changed it all to singletons managed by MEF now.
it used to be so that it required request data to first initialize, but turns out that's not needed